pkg/pgclient/config.go

```diff
@@ -135,70 +139,57 @@ func (cfg *Config) GetConnectionStr() string {
 	return cfg.DbUri
 }
 
-func (cfg *Config) GetNumConnections() (min int, max int, numCopiers int, err error) {
-	maxProcs := runtime.GOMAXPROCS(-1)
-	if cfg.WriteConnections < 0 {
-		return 0, 0, 0, fmt.Errorf("invalid number of writer connections %v, cannot be a negative number", cfg.WriteConnections)
+// GetPoolSize returns the max pool size based on the max_connections allowed by database and the defaultFraction.
+// Arg inputPoolSize is the pool size provided in the CLI flag.
+func (cfg *Config) GetPoolSize(poolName string, defaultFraction float64, inputPoolSize int) (int, error) {
+	if inputPoolSize != defaultPoolSize && inputPoolSize < MinPoolSize {
+		return 0, fmt.Errorf("%s pool size canot be less than %d: received %d", poolName, MinPoolSize, inputPoolSize)
 	}
 
-	conn, err := pgx.Connect(context.Background(), cfg.GetConnectionStr())
+	maxConns, err := cfg.maxConn()
 	if err != nil {
-		return 0, 0, 0, err
+		return 0, fmt.Errorf("max connections: %w", err)
+	}
+	if inputPoolSize > maxConns {
+		return 0, fmt.Errorf("%s pool size canot be greater than the 'max_connections' allowed by the database", poolName)
 	}
-	defer func() { _ = conn.Close(context.Background()) }()
 
-	max = cfg.MaxConnections
-	if max < 1 {
-		var maxStr string
-		row := conn.QueryRow(context.Background(), "SHOW max_connections")
-		err = row.Scan(&maxStr)
-		if err != nil {
-			return 0, 0, 0, err
-		}
-		max, err = strconv.Atoi(maxStr)
-		if err != nil {
-			log.Warn("err", err, "msg", "invalid value from postgres max_connections")
-			max = 100
-		}
+	// For the default case, we need to take up defaultFraction of allowed connections.
+	poolSize := float64(maxConns) * defaultFraction
+	if cfg.UsesHA {
+		poolSize /= 2
+	}
+	return int(poolSize), nil
+}
 
-		//In HA setups
-		if cfg.UsesHA {
-			max = max / 2
-		}
+func (cfg *Config) maxConn() (int, error) {
+	conn, err := pgx.Connect(context.Background(), cfg.GetConnectionStr())
```

Comment:
In `GetNumCopiers` you call `defer func() { _ = conn.Close(context.Background()) }()` but not here.

Reply:
Thank you for catching. This might fix the CI that is complaining for some unclosed connection.